networking/src/Network/Broadcast/OutboundQueue.hs

```diff
@@ -326,19 +348,41 @@ inFlightWithPrec nid prec = inFlightTo nid . at prec . anon 0 (== 0)
 
 -- | Given an update function and a `Packet`, set the `InFlight` to the new value
 -- calculated by `f`. In case no match can be found, this function is effectively a noop.
-setInFlightFor :: forall m msg nid a. (MonadIO m, Ord nid)
+setInFlightFor :: forall msg nid a. (Ord nid)
                => Packet msg nid a
                -> (Int -> Int)
                -> MVar (InFlight nid)
-               -> m ()
-setInFlightFor Packet{..} f var = liftIO $ modifyMVar_ var (return . update)
+               -> IO ()
+setInFlightFor Packet{..} f var = modifyMVar_ var (return . update)
   where
     update :: InFlight nid -> InFlight nid
     update = Map.adjust updateInnerMap packetDestId
 
     updateInnerMap :: Map Precedence Int -> Map Precedence Int
     updateInnerMap = Map.adjust f packetPrec
 
+-- | Abstracts logging.
+newtype Trace m s = Trace
```

Comment:
I was shy to put too much of an emphasis on this, because I worry that people are extremely fond of monad-transformer-based logging (that seems to be the way that everybody does it). Lately I've come to not like it so much. Why should I not be able to use `IO`, just because of logging?
Anyway, yes this does deserve a better place. Do bring it out to its own module and use it in your NTP work if you think it would be useful.

Reply:
+1 I like it, it's a simple abstraction that let both camps to live alongside and indeed logging shouldn't decide on the monad one wants to use.

Reply:
And if one provides profanctor instance instead of the covariant one, one could also lift the monad.

Reply:
Profunctor wouldn't be appropriate, since the monadic term should always have a `()`. And a profunctor-like thing to change the monad would need a different kind:

```
class WeirdProfunctor (f :: * -> (* -> *) -> *) where
  weirdRmap :: (forall x . m x -> n x) -> f a m -> f a n
  normalLmap :: (b -> a) -> f a m -> f b m
```

Reply:
I know; this is what I had in mind. The `weiredRmap` is just application of a natural transformation. It's kind of a skewed profunctor, rather than just a functor: `C^op x C -> Set`, a functor of this sort: `A^op x B -> Set`